src/Foundatio.AWS/Queues/SQSQueue.cs

```diff
@@ -0,0 +1,269 @@
+            _readQueueTimeout = readQueueTimeout ?? TimeSpan.FromSeconds(2);
+
+            _client = new Lazy<AmazonSQSClient>(() => new AmazonSQSClient(_credentials, _regionEndpoint));
+        }
+
+        protected override async Task EnsureQueueCreatedAsync(CancellationToken cancellationToken = new CancellationToken()) {
+            if (!String.IsNullOrEmpty(_queueUrl))
+                return;
+
+            using (await _lock.LockAsync(cancellationToken).ConfigureAwait(false)) {
+                if (!String.IsNullOrEmpty(_queueUrl))
+                    return;
+
+                // per sqs documentation, create will get existing or create new queue
+                var createRequest = new CreateQueueRequest { QueueName = _queueName };
+                var createResponse = await _client.Value.CreateQueueAsync(createRequest, cancellationToken).ConfigureAwait(false);
+
+                _queueUrl = createResponse.QueueUrl;
+            }
+        }
+
+        protected override async Task<string> EnqueueImplAsync(T data) {
+            if (!await OnEnqueuingAsync(data).ConfigureAwait(false))
+                return null;
+
+            Interlocked.Increment(ref _enqueuedCount);
+            var message = new SendMessageRequest {
+                QueueUrl = _queueUrl,
+                MessageBody = await _serializer.SerializeToStringAsync(data).ConfigureAwait(false),
+            };
+
+            var response = await _client.Value.SendMessageAsync(message).ConfigureAwait(false);
+
+            var entry = new QueueEntry<T>(response.MessageId, data, this, SystemClock.UtcNow, 0);
+            await OnEnqueuedAsync(entry).ConfigureAwait(false);
+
+            return response.MessageId;
+        }
+
+        protected override async Task<IQueueEntry<T>> DequeueImplAsync(CancellationToken cancellationToken) {
+            var visibilityTimeout = Convert.ToInt32(_workItemTimeout.TotalSeconds);
+
+
+            // sqs doesn't support already canceled token, change timeout and token for sqs pattern
+            var waitTimeout = cancellationToken.IsCancellationRequested
+                ? 0
+                : Convert.ToInt32(_readQueueTimeout.TotalSeconds);
+
+            var cancel = cancellationToken.IsCancellationRequested
+                ? CancellationToken.None
+                : cancellationToken;
+
+            var request = new ReceiveMessageRequest {
+                QueueUrl = _queueUrl,
+                MaxNumberOfMessages = 1,
+                VisibilityTimeout = visibilityTimeout,
+                WaitTimeSeconds = waitTimeout,
+                AttributeNames = new List<string> { "ApproximateReceiveCount", "SentTimestamp" }
+            };
+
+            ReceiveMessageResponse response = null;
+
+            try {
+                response = await _client.Value.ReceiveMessageAsync(request, cancel).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException) {
```

Comment:
@StephenCleary recommended we catch the base `OperationCanceledException` exception for all cancellations.